lib/FileService.php

```diff
@@ -133,6 +133,55 @@ public function getFileHandle(int $fileId, string $ownerUID, string $userUID): ?
 		return null;
 	}
 
+	public function newFile(string $path, string $ownerUID, string $editorUID): ?File {
+		if ($path === '') {
+			return null;
+		}
+
+		if ($editorUID) {
+			$user = $this->userManager->get($editorUID);
+			if (!$user) {
+				$this->logger->warning('newFile(): No such user', ['app' => 'richdocuments']);
+				return null;
+			}
+
+			// Make sure editor session is opened for registering activity over file handle
+			$this->logger->debug('newFile(): Register session as ' . $editorUID, ['app' => 'richdocuments']);
+			if (!$this->appConfig->encryptionEnabled()) {
+				// Set session for a user
+				$this->userSession->setUser($user);
+			} elseif ($this->appConfig->masterEncryptionEnabled()) {
+				// With master encryption, decryption is based on master key (no user password needed)
+				// make sure audit/activity is triggered for editor session
+				$this->userSession->setUser($user);
+
+				// emit login event to allow decryption of files via master key
+				$afterEvent = new GenericEvent(null, ['loginType' => 'password', 'user' => $user, 'uid' => $editorUID, 'password' => '']);
+				$this->eventDispatcher->dispatch($afterEvent, 'user.afterlogin');
+			} else {
+				return null;
+			}
+		} else {
+			return null;
+		}
+
+		// Setup FS of original file file-handle to be able to generate
+		// file versions and write files with user session set for editor
+		// take 1st mount as here we access original version of the file from owner
+		$this->setupFS($ownerUID);
+
+		// create the folder first
+		if (!$this->rootFolder->nodeExists(\dirname($path))) {
+			$this->rootFolder->newFolder(\dirname($path));
+		}
+
+		// create a unique new file
+		$path = $this->rootFolder->getNonExistingName($path);
```

Comment:
I guess race conditions aren't worthy to check in this case. Any dangerous thing that could happen an case of a race condition?

Reply:
worked for many years, so hopefuly would work still . this is just move of a code to the dedicated class.